[Woo POS][Local Catalog] Add incremental sync triggers

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

WOOMOB-1098
Done based on the project test plan: pdfdoF-7RD-p2

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Add incremental sync triggers:
- Pull to refresh
- Payment completion
- POS opened
- App opened

## Solution

### Pull to Refresh

For PTR, I needed to update `PointOfSaleObservableItemsController` to support `refreshItems` method. Within it, I added `coordinator.performIncrementalSync` call. However, the error handling was more trickier part. I had to introduce `refreshState` to track when the refresh happens or fails.

`ChildItemList` was also updating every time anything was changing in observable, therefore, the `.refreshable` would get canceled immediately after `refreshState` changed. I had to wrap it in another task to make sure it doesn't get canceled automatically. 

`refreshState` needs to be included in multiple checks to handle error and empty list cases, to make sure the refresh is triggered at the right times.

### Payment Completion

For payment completion, I'm using `Observation` framework observation tracking to check for cash and card payment success to trigger incremental sync

### POS Opened

The most straightforward place was to trigger sync within aggregate model loading.

### App Opened

It was already implemented, but I updated `performSmartSync` to not do incremental sync if it has happened within an hour.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Launch POS
1. Confirm in logs that incremental sync is done

### PTR on Parent and Child view
1. Do pull to refresh on parent and child views
2. Check logs, confirm incremental sync happens
3. Make a change on wp-admin
4. Do PTR, confirm changes appear (note incremental sync for variations has issues WOOMOB-1546)

### PTR + Errors + Recovery
1. Block connection before PTR
2. PTR, confirm inline error appears
3. Re-enable connection
4. PTR / Tap Retry
5. Confirm the incremental sync is done

### Payment completion
- Complete cash payment, confirm in logs that incremental sync is done
- Complete card payment, confirm in logs that incremental sync is done

### App open + incremental sync skipped if 1h not passed
1. Do pull to refresh
2. Go to background
3. Come back to foreground
4. Check logs, confirm the incremental sync is skipped

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad Air 26 simulator, new and established sites.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Incremental sync on: Open POS, PTR on parent and child, payment completion, skipped on app launch

https://github.com/user-attachments/assets/6037e3ac-0638-42a6-881d-a5e468273e18

### Empty State

https://github.com/user-attachments/assets/bfa88b4f-5d19-4012-94ed-5a02fe972e33

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Skip incremental sync if one hour hasn't passed in performSmartSync

staskus committed:
Perform incremental POS catalog sync on pull to refresh

staskus committed:
Throw errors from missing paths in POSCatalogSyncCoordinator

Catalog size limit check can fail due to network issues and get unreported up the chain.

Throw an error when that happens.

staskus committed:
Remove unnecessary tests from GRDBObservableDataSourceTests

staskus committed:
Add error handling to PointOfSaleObservableItemsController

Set refreshError when syncing fails. Most often it will be an inline error, showed together with the content.

Also, if error is set, refresh items when loadItems is called, which happens when retry is tapped on the error.

staskus committed:
Add a helper for continuously tracking observations

staskus committed:
Perform incremental sync after a successful payment on POS

staskus committed:
Verify that incremental sync is triggered on successful payment

staskus committed:
Perform incremental sync when POS launches

staskus committed:
Ensure ChildItemList refresh task is not canceled even if the view is redrawn

staskus committed:
Add a RefreshState to handle loading and item list view cases

staskus committed:
Make withObservationTracking internal

staskus committed:
Merge branch 'trunk' into woomob-1098-woo-poslocal-catalog-add-incremental-sync-triggers-pull-to

joshheald commented on Modules/Sources/PointOfSale/Presentation/Item Selector/ChildItemList.swift:
```diff
@@ -78,7 +78,9 @@ private extension ChildItemList {
-                await itemsController.refreshItems(base: node)
+                await Task {
+                    await itemsController.refreshItems(base: node)
+                }.value
```
I thought that this is only needed because we set the state to `loading` on a refresh, which isn't really needed – the refreshing spinner that iOS puts up for us is the indicator that something's happening. However, removing the task shows that some `explicitlyCancelled` errors are coming through in that case.

I remember writing some code to swallow these – they're not real errors, it's an AlamoFire thing when a previous request gets cancelled, I think. Perhaps we need the same in the incremental sync service somewhere? With that, and removing the call to always set a loading state, I think we could get rid of the task.

staskus replied:
I noticed it happening even before I had a loading state, and only had an error state. It was enough to set the error to nil before the incremental sync to cancel it.

In short, if I made any view updates while the refreshItems was still ongoing, the SwiftUI would kill the refreshable task, which caused an explicitlyCancelled error to happen. Even if we swallow the error not to be displayed, the refresh task remains canceled. 

I think the problem may be with the view structure, since it doesn't happen on the ItemListView. However, I wasn't able to find a structure that wouldn't produce this cancellation...

joshheald replied:
Can you reproduce it with the existing item controller? If not, perhaps that rules out the view structure?

> In short, if I made any view updates while the refreshItems was still ongoing, the SwiftUI would kill the refreshable task, which caused an explicitlyCancelled error to happen. Even if we swallow the error not to be displayed, the refresh task remains canceled.

Yes, that's unavoidable unfortunately. I'm confusing two memories, we had explicitly cancelled errors from elsewhere for something else. The key is that we can't change the view while we're refreshing.

staskus replied:
> Can you reproduce it with the existing item controller? If not, perhaps that rules out the view structure?

I'll try!

joshheald commented on Modules/Sources/PointOfSale/Presentation/Item Selector/ChildItemList.swift:
```diff
@@ -78,7 +78,9 @@ private extension ChildItemList {
-                await itemsController.refreshItems(base: node)
+                await Task {
+                    await itemsController.refreshItems(base: node)
+                }.value
```
```suggestion
                await itemsController.refreshItems(base: node)
```

joshheald commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -108,12 +149,21 @@ private extension PointOfSaleObservableItemsController {
-        // Loading state - preserve existing items
-        if dataSource.isLoadingProducts {
+        // Loading state - preserve existing items (both for data loading and refresh)
+        if dataSource.isLoadingProducts || refreshState == .loading {
```
```suggestion
        // Loading state - preserve existing items (only use for data loading, refresh has a PTR spinner)
        if dataSource.isLoadingProducts && refreshState != .loading {
```

We shouldn't show the loading cell if we're refreshing – the PTR indicator is sufficient, and changing the state breaks the refresh task anyway.

staskus replied:
I implemented it to support the error "retry" case, but we could live without it. It's a shame there's no way to manually show a PTR loading animation. I remember trying to replicate it without success.

joshheald commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -139,12 +189,21 @@ private extension PointOfSaleObservableItemsController {
-        // Loading state - preserve existing items
-        if dataSource.isLoadingVariations {
+        // Loading state - preserve existing items (both for data loading and refresh)
+        if dataSource.isLoadingVariations || refreshState == .loading {
```
```suggestion
        // Loading state - preserve existing items (only use for data loading, refresh has a PTR spinner)
        if dataSource.isLoadingProducts && refreshState != .loading {
```

joshheald commented on Modules/Sources/PointOfSale/Controllers/PointOfSaleObservableItemsController.swift:
```diff
@@ -6,19 +6,24 @@ import protocol Yosemite.POSObservableDataSourceProtocol
 
     // Track current parent for variation state mapping
     private var currentParentItem: POSItem?
+    private var refreshState: RefreshState = .idle
```
What do you think about this being shared between root and variations? It feels like an opportunity to leak state between views by accident...

staskus replied:
Well, I made it shared since the incremental sync itself is shared. There's no separate sync for specific parent views. How do you think I should structure it?

joshheald replied:
Ah yes, I see. Makes sense.

The one niggle I have is that this can be out of sync with the view state, because Apple's code's in control of which PTR indicator is shown. If you PTR on a variation screen, then go back to a parent screen and scroll down to load more items, we'd be loading and refreshing at the same time. Which is kind of OK... but if we use the refresh state to decide to show a loading indicator, we could finish one load without finishing the incremental sync, and still show a loading indicator for the refresh.

If we remove the loading cell for refresh tasks it's less of an issue... and we deliberately don't show that for PTR in the existing implementation, as the spinner is enough.

joshheald reviewed (approved):
> However, the error handling was more trickier part. I had to introduce refreshState to track when the refresh happens or fails.

Would this be needed if we didn't have to wrap the refresh task in a `Task` block? Since it's not required in the existing implementation, I assume not?

It seems like it would be a lot simpler if we didn't have that extra state...

Potentially it can be removed. The important thing is not to change the view while there's a refresh happening. Perhaps that's unavoidable while we're observing the underlying database... but I don't _think_ it should be changing except when the refresh task completes... and maybe that would be fine?

I've not had time to dig deep, can sync on it tomorrow if you like. We did face similar issues with the original implementation and managed to fix them all so that we didn't have to swallow the errors with that Task block.

There's a few questions and suggestions in line too.

It does work as expected and test well, so I'm cautiously approving it, but it'd be really good to explore further whether we can remove refreshState before merging. Let me know what you think...

staskus committed:
Merge branch 'trunk' into woomob-1098-woo-poslocal-catalog-add-incremental-sync-triggers-pull-to

staskus commented:
@joshheald, thanks for testing!

> Would this be needed if we didn't have to wrap the refresh task in a Task block? Since it's not required in the existing implementation, I assume not?
> It seems like it would be a lot simpler if we didn't have that extra state...
> Potentially it can be removed. The important thing is not to change the view while there's a refresh happening. Perhaps that's unavoidable while we're observing the underlying database... but I don't think it should be changing except when the refresh task completes... and maybe that would be fine?

True, we could avoid that if we wouldn't be able to set `error` to `nil` until the refresh is complete. So when the user taps "Retry" on an error, there would be no indication of progress. I could try changing the state separately from the view side, that would be a bit of a hack.

The best thing would be to figure out why SwiftUI redraws the ChildList and kills the refreshable task when the error changes but it doesn't do that for the ItemListView.

> It does work as expected and test well, so I'm cautiously approving it, but it'd be really good to explore further whether we can remove refreshState before merging. Let me know what you think...

I'm not sure I can remove `refreshState` entirely, we still need to have some state to display errors.

joshheald commented:
> True, we could avoid that if we wouldn't be able to set `error` to `nil` until the refresh is complete. So when the user taps "Retry" on an error, there would be no indication of progress.

Could we do that and indicate progress for the retry using a different indicator? Perhaps a spinner on the retry button? I think the UX makes sense – you PTR, it fails, we show an error. You then tap a button to retry, it's not a PTR any more, so an indicator on the button is perhaps clearer.

staskus commented:
> Could we do that and indicate progress for the retry using a different indicator? Perhaps a spinner on the retry button? I think the UX makes sense – you PTR, it fails, we show an error. You then tap a button to retry, it's not a PTR any more, so an indicator on the button is perhaps clearer.

@joshheald thanks, I'll try that. I think we have a button loading state developed.

staskus commented:
@joshheald I also remembered one more case for the loading state. When we have an empty view and we want to show a loading view when the refresh is tapped. However, we could only affect this particular state.

joshheald commented:
> @joshheald I also remembered one more case for the loading state. When we have an empty view and we want to show a loading view when the refresh is tapped. However, we could only affect this particular state.

Yeah. I think the key is: don't change the state when doing a Pull to Refresh, let Apple handle it.

staskus committed:
Add a loading state for an outlined button style

staskus committed:
Set loading state to POSListInlineErrorView

staskus committed:
Only refresh items list if there are no products

Additionally, remove the business logic duplications from the controller

staskus committed:
Ignore requestCancelled failures

They can happen when PTR is started in one view and we move to another. No need to show an error.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/258">23.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Remove unused code

staskus committed:
Remove unnecessary "Observation+WithObservationTrackingOf" file

staskus committed:
Compare two POSCatalogSyncError in PointOfSaleObservableItemsController

staskus committed:
Remove a task wrapper around refreshable to avoid cancelation

This hack was required to support child view changes while refresh is happening. However, we avoid changing the state now while refresh is happening so it's no longer necessary.

staskus commented:
- Removed the `Task {}` wrapped in the refreshable task
- Added a retry animation on an inline error
- Made error comparison more robust
- Only showing loading from error/empty if there are no items
- Ignoring cancellation errors as in other controllers

https://github.com/user-attachments/assets/4195c314-7119-48d7-a8d0-dc6e63dc11b0

staskus commented:
@joshheald 

> It does work as expected and test well, so I'm cautiously approving it, but it'd be really good to explore further whether we can remove refreshState before merging. Let me know what you think...

I wasn't able to remove refreshState, so I'm not forcing a merge for now, especially if there are some unexplored ways 🤔

joshheald commented:
OK, I'll take it over and try the other ways early next week. I'll probably end up merging as is, but at least we'll be sure

staskus commented:
Thanks! @joshheald. I wanted to get it over the line but better be safe with this core functionality.

joshheald committed:
Merge branch 'trunk' into woomob-1098-woo-poslocal-catalog-add-incremental-sync-triggers-pull-to

joshheald commented:
I've tested this further, and will merge it as it is. We can improve it at a later date, if we need to. It works well at the moment, and the theory of avoiding this extra state isn't worth the time right now, it's pretty well encapsulated.

https://linear.app/a8c/issue/WOOMOB-1613/local-catalog-consider-removing-refreshstate-from
